scripts/typemock/src/times.ts

```diff
@@ -0,0 +1,120 @@
+    }
+
+    /**
+     * Expects that an action was executed at least once.
+     * @returns A new `Times` instance.
+     */
+    public static atLeastOnce(): Times {
+        return this._atLeastOnce || (this._atLeastOnce = new Times(1, Number.MAX_SAFE_INTEGER, `at least once`));
+    }
+
+    /**
+     * Expects that an action was executed at least the specified number of times.
+     * @param count The number of times.
+     * @returns A new `Times` instance.
+     */
+    public static atLeast(count: number): Times {
+        return new Times(count, Number.MAX_SAFE_INTEGER, `at least ${count} time(s)`);
+    }
+
+    /**
+     * Expects that an action was executed exactly the specified number of times.
+     * @param count The number of times.
+     * @returns A new `Times` instance.
+     */
+    public static exactly(count: number): Times {
+        return new Times(count, count, `exactly ${count} time(s)`);
+    }
+
+    /**
+     * Expects that an action was executed at most the specified number of times.
+     * @param count The number of times.
+     * @returns A new `Times` instance.
+     */
+    public static atMost(count: number): Times {
+        return new Times(0, count, `at most ${count} time(s)`);
+    }
+
+    /**
+     * Expects that an action was executed at most once.
+     * @returns A new `Times` instance.
+     */
+    public static atMostOnce(): Times {
+        return this._atMostOnce || (this._atMostOnce = new Times(0, 1, `at most once`));
+    }
+
+    /**
+     * Expects that an action was executed between a range of times, inclusive.
+     * @param min The minimum number of times, inclusive.
+     * @param max The maximum number of times, inclusive.
+     * @returns A new `Times` instance.
+     */
+    public static between(min: number, max: number): Times {
+        return new Times(min, max, `between ${min} and ${max} time(s)`);
+    }
+
+    /**
+     * Validates the number of times an action was executed.
+     * @param count The number of times the action was executed.
+     * @returns `true` if the provided count was valid; otherwise, `false`.
+     */
+    public validate(count: number): boolean {
+        if (count < this._min) return false;
+        if (count > this._max) return false;
+        return true;
+    }
+
+    /**
+     * Checks the number of times an action was executed, throwing an error if the count was not valid.
+     * @param count The number of times the action was executed.
+     * @param message The message to use to begin the check.
+     */
+    public check(count: number, message: string): void {
+        if (!this.validate(count)) {
+            const expectedMessage = this._message === `never`
+                ? `Expected to never be executed.`
+                : `Expected to be executed ${this._message}.`;
+            throw new Error(`${message}\n${expectedMessage} Actually executed ${count} time(s).`);

```

Comment:
Minor enhancement: You can use errors of the (richer) shape of an [assertion error](https://www.npmjs.com/package/assertion-error) to get better output from mocha for actual/expected results. AFAIK, it's [just the addition of an `expected` and an `actual` property](https://github.com/mochajs/mocha/blob/master/lib/reporters/base.js#L446) onto an error.

Reply:
That makes sense when comparing two literal values, but the comparisons here are less relevant. What would you put for `expected` and `actual` for `Times.none().validate(1)`? If you used `expected = "Expected to never be executed"` and `actual = "Actually executed 1 time(s)."` what would be the added value? If you used a diff, then it would be a string diff between the two strings which isn't all that useful.

Reply:
The [message should also be preserved](https://github.com/mochajs/mocha/blob/master/lib/reporters/base.js#L216), so if you set `expected` to the expected range of executions, and `actual` to the number of executions, you should get a colored output for the ranges, while the context message could be something like `${message}: ${expectedMessage}, but actually executed ${count}.` (though `expectedMessage` needs to starts with a lowercase `e`-`expected` for that to work)

Reply:
As I said, a diff wouldn't make sense, so `expected` and `actual` aren't as useful. I don't think it's worth the added complexity.

Reply:
`¯\_(ツ)_/¯` I just usually find spotting the differences easier with the aid of the differing coloring the expected/actual split provides. It's not that important.

Reply:
I think diffing makes sense when you are comparing literals (like "aaa" and "aAa") where you want to know *where* the difference occurred. That doesn't apply to this case, because the *where* isn't the text, its the place in the code where the mock/spy was executed.